File: w1/Ex5.py
# This is the better version

def threeSum(arr):
    arr.sort()
    ans = []
    if len(arr)<3:
        return []
    for i,j in enumerate(arr):
        l = i+1
        r = len(arr)-1
        while l < r:
            if i>0 and arr[i-1] == j:
                break
            summ = j+arr[l]+arr[r]
            if summ<0:
                l+=1
            elif summ>0:
                r-=1
            else:
                ans.append([arr[l],arr[r],j])
                r-=1
    return ans

arr = [-1,-2,0,1,2,-3,4,-3]
print(threeSum(arr))

File: w1/Ex5.java
import java.util.Arrays;

// The python version is better

public class Ex5{

    public static void main(String[] args){
        int[] arr = {-3,-2,-1,0,1,2};
        Arrays.sort(arr);
        
        for (int i=0;i<arr.length;i++){
            for (int j=i+1;j<(i+arr.length);j++){
                int betterj = j%arr.length;
                int p = (j+1) % arr.length;
                if (p == i){
                    p = (p+1)%arr.length;
                }

                // System.out.print(i);
                // System.out.print(betterj);
                // System.out.println(p);

                if (-arr[i]-arr[betterj] == arr[p]){
                    System.out.print(arr[i]);
                    System.out.print(arr[betterj]);
                    System.out.print(arr[p]);
                    System.out.print(" ");
                }
                
            }
        }
    }
}
